My cart Page

Chandrika-Rani-Infrrd opened the pull request:
added:routing to product-list to navigate to my cart
updated :product list spec file with few test case
updated:mycart file to get and display selected products
added:quantity increment and decrement,delete,total amount to my cart page
updated :mycart spec file with few test cases
updated:product json with id and quantity property

Chandrika-Rani-Infrrd committed:
My cart Page
added:routing to product-list to navigate to my cart
updated :product list spec file with few test case
updated:mycart file to get and display selected products
added:quantity increment and decrement,delete,total amount to my cart page
updated :mycart spec file with few test cases
updated:product json with id and quantity property

lloydaf commented on src/app/mycart/my-cart.component.html:
```diff
@@ -1 +1,24 @@
+        <button class="delete" (click)="productToDelete(item.id)">Delete</button>
+    </div>
+</div> 
+<div class="total">Total Amount : {{sum()}}</div>
```
You shouldn't call functions from the template unless they are based on an event, like click, scroll etc. Template functions will get triggered on every change detection cycle. Use a class property instead.

lloydaf commented on src/app/mycart/my-cart.component.html:
```diff
@@ -1 +1,24 @@
+            <button (click)="increment(item)" >+</button>
+        </div>
+        <div><b>Price : </b>{{item.price}}</div>
+        <div><b>Total Price : </b>{{total(item.quantity,item.price)}}</div>
```
Don't use a function, use a property or a pipe instead

lloydaf replied:
Reason explained below 👇

lloydaf commented on src/app/mycart/my-cart.component.spec.ts:
```diff
@@ -22,4 +31,57 @@ describe('MycartComponent', () => {
+  fixture.detectChanges(); 
+
+  router.navigate(['product-list']).then(() => {
+    expect(location.path()).toBe('/product-list'); 
```
💯

lloydaf commented on src/app/mycart/my-cart.component.spec.ts:
```diff
@@ -22,4 +31,57 @@ describe('MycartComponent', () => {
+  
+  it("should have getProducts method",()=>{
+    const service:ProductsService=TestBed.get(ProductsService)
+    expect(service.getProducts).toBeTruthy();
```
💯

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+  productsList:any=[];
+  myCart:any=[];
+  cartValues=[];
+  totalAmount:number=0;
```
you needn't really specify the type here. Javascript can imply the type from the value

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+
+      for(let i=0;i<this.productsList.length;i++){
+        if(this.productsList[i].name===this.name && 
+          !this.cartValues.map(item=>item.name).includes(this.name)){
```
You shouldn't use a loop within a loop. `O(n^2)` time complexity.

lloydaf replied:
To resolve this issue, make ProductsList a set or a map. You don't have to do it that way, just a suggestion, upto you

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+          !this.cartValues.map(item=>item.name).includes(this.name)){
+            this.cartValues.push(this.productsList[i]);
+            console.log("after push",this.productsList[i]);
+            localStorage.setItem('cartValue',JSON.stringify(this.cartValues));
```
you shouldn't really do this within every iteration. Expensive operation right?

Chandrika-Rani-Infrrd replied:
can i create a service and then use only cartvalues in mycart file?

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+            this.cartValues.push(this.productsList[i]);
+            console.log("after push",this.productsList[i]);
+            localStorage.setItem('cartValue',JSON.stringify(this.cartValues));
+            let cart=JSON.parse(localStorage.getItem('cartValue'));
```
You shouldn't probably do this either

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+
+  productToDelete(value){
+    console.log("item"+value);
+    this.cartValues.splice(this.cartValues.findIndex(item=>item.id==value),1);
```
you can just use a filter instead, than using findIndex within splice. Try to keep your code as functional as possible, with minimum mutations. you can read about it here https://medium.com/@cristiansalcescu/learn-immutability-with-javascript-6a67e4a48d7f

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+  }
+
+  total(quantity,price):number{
+    let total=(quantity*price);
```
you don't need a function for this

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,82 @@
+
+  sum():number{
+    let sum=0;
+    for(let i in this.cartValues){
```
you can use `.reduce` instead of a for loop here

Chandrika-Rani-Infrrd committed:
My Cart page
added:pipes for mycart to get total price and total amount
updated:mycart by adding/removing few code lines
added:few spec file with test cases

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,80 @@
+              private productService:ProductsService){}
 
   ngOnInit(): void {
+    if(localStorage.getItem('deleteditem').length>0){
```
use a variable instead of a hard coded string value

lloydaf commented on src/app/mycart/over-all-total.pipe.ts:
```diff
@@ -0,0 +1,18 @@
+  }
+
+  overAllTotal(sum,products){
+    sum=products.reduce((acc,item)=>acc+(item.quantity*item.price),0);
```
what is the purpose of passing sum as an argument if you are going to overwrite the value?

Chandrika-Rani-Infrrd committed:
My Cart Page
updated :productlist file to send selected product to cart service
updated:productlist spec with few test cases
udated :cart service to send selected  product to mycart component
updated:mycart component with addition/deletion of code
added:delete component where it performs delete operation
updated:mycart to call that delete coponent method

lloydaf commented on src/app/app-routing.module.ts:
```diff
@@ -4,16 +4,15 @@ import { Routes, RouterModule } from '@angular/router';
+  { path: '', redirectTo: 'product-list', pathMatch: 'full' },
+  { path: 'product-list', component: ProductListComponent },
+  { path: 'mycart', component: MyCartComponent },
+  { path: 'mycart/:name', component: MyCartComponent },
```
remove this `:name` parameter

lloydaf commented on src/app/mycart/my-cart.component.spec.ts:
```diff
@@ -22,4 +40,60 @@ describe('MycartComponent', () => {
   });
+
+  it('create an instance for pipe', () => {
+    const pipe = new OverAllTotalPipe();
```
💯

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,60 @@
 })
 export class MyCartComponent implements OnInit {
+  myCartValues = [];
+  name: string;
```
you don't need this

lloydaf commented on src/app/mycart/my-cart.component.ts:
```diff
@@ -1,15 +1,60 @@
+  ) {}
 
   ngOnInit(): void {
+    this.name = this.route.snapshot.paramMap.get('name');
```
don't need this

lloydaf commented on src/app/mycart/my-cart.service.ts:
```diff
@@ -0,0 +1,17 @@
+export class MyCartService {
+
+  private selectedItem:any;
+  
```
addItemToCart(),
removeItemFromCart(),
increment, decrement

Chandrika-Rani-Infrrd committed:
My Cart Page
updated:cartservice file with add,remove,increment,decrement methods
updated:mycart component which calls all method of cartservic file
